src/hooks/wallets/useInitWeb3.ts

```diff
@@ -6,29 +6,33 @@ import { createWeb3, createWeb3ReadOnly, setWeb3, setWeb3ReadOnly } from '@/hook
 import { useAppSelector } from '@/store'
 import { selectRpc } from '@/store/settingsSlice'
 
+const READONLY_WAIT = 1000
+
 export const useInitWeb3 = () => {
   const chain = useCurrentChain()
-  const chainId = chain?.chainId
-  const rpcUri = chain?.rpcUri
   const wallet = useWallet()
   const customRpc = useAppSelector(selectRpc)
   const customRpcUrl = chain ? customRpc?.[chain.chainId] : undefined
 
   useEffect(() => {
-    if (wallet && wallet.chainId === chainId) {
+    if (!chain) return
+
+    if (wallet) {
       const web3 = createWeb3(wallet.provider)
       setWeb3(web3)
-    } else {
-      setWeb3(undefined)
-    }
-  }, [wallet, chainId])
 
-  useEffect(() => {
-    if (!rpcUri) {
-      setWeb3ReadOnly(undefined)
-      return
+      if (wallet.chainId === chain.chainId) {
+        setWeb3ReadOnly(web3)
+        return
+      }
     }
-    const web3ReadOnly = createWeb3ReadOnly(rpcUri, customRpcUrl)
-    setWeb3ReadOnly(web3ReadOnly)
-  }, [rpcUri, customRpcUrl])
+
+    // Wait for wallet to be connected
+    const timeout = setTimeout(() => {
+      const web3ReadOnly = createWeb3ReadOnly(chain.rpcUri, customRpcUrl)
+      setWeb3ReadOnly(web3ReadOnly)
+    }, READONLY_WAIT)
```

Comment:
Why do we need to wait if there is no wallet connected?

Reply:
Becuase the wallet takes some time to connect, we don't want to initialize a read-only provider (e.g. Infura) and make the first few RPC calls before a connected wallet kicks in.